File: BackEnd/JavaScript/expressshop_prismaorm/src/image/image.service.js
import crypto from "crypto";

import { badRequest, notFound } from "../utils/http-error.js";
import {
  createImageAssetWithVariants,
  deleteImageAssetWithVariants,
  findImageAssetById,
  findImageVariants,
} from "./image.repository.js";

const ALLOWED_CATEGORIES = new Set(["product", "community", "support", "seller"]);
const ALLOWED_MIME_TYPES = new Set(["image/jpeg", "image/png", "image/webp", "image/gif"]);

function getExtension(filename) {
  const parts = String(filename ?? "").split(".");
  return parts.length > 1 ? parts.at(-1).toLowerCase() : "bin";
}

function createStoredToken() {
  return `${Date.now()}-${crypto.randomBytes(4).toString("hex")}`;
}

function toVariantDto(item) {
  return {
    id: item.id,
    type: item.type,
    url: item.url,
    width: item.width,
    height: item.height,
    format: item.format,
    size: item.size,
    createdAt: item.createdAt,
  };
}

function buildVariants(token, size) {
  return [
    {
      type: "THUMBNAIL",
      url: `/uploads/thumb/${token}.webp`,
      format: "WEBP",
      width: 200,
      height: 200,
      size: Math.max(1, Math.floor(size * 0.12)),
    },
    {
      type: "MEDIUM",
      url: `/uploads/medium/${token}.webp`,
      format: "WEBP",
      width: 600,
      height: 600,
      size: Math.max(1, Math.floor(size * 0.35)),
    },
    {
      type: "LARGE",
      url: `/uploads/large/${token}.webp`,
      format: "WEBP",
      width: 1200,
      height: 1200,
      size: Math.max(1, Math.floor(size * 0.65)),
    },
  ];
}

export async function uploadImage(userId, file, body) {
  if (!file) {
    throw badRequest("file is required");
  }
  if (!ALLOWED_MIME_TYPES.has(file.mimetype)) {
    throw badRequest("file type is not allowed");
  }
  if (!ALLOWED_CATEGORIES.has(body?.category)) {
    throw badRequest("category is invalid");
  }

  const token = createStoredToken();
  const storedFilename = `${token}.${getExtension(file.originalname)}`;
  const image = await createImageAssetWithVariants(
    {
      uploadedByUserId: Number(userId),
      originalFilename: file.originalname,
      storedFilename,
      originalUrl: `/uploads/original/${storedFilename}`,
      mimeType: file.mimetype,
      size: Number(file.size ?? 0),
      category: body.category,
      processingStatus: "PROCESSING",
    },
    buildVariants(token, Number(file.size ?? 0)),
  );

  return {
    id: image.id,
    originalUrl: image.originalUrl,
    variants: image.variants.map(toVariantDto),
    processingStatus: image.processingStatus,
  };
}

export async function uploadImageWithCategory(userId, file, category) {
  return uploadImage(userId, file, { category });
}

export async function getImageVariants(imageId) {
  const image = await findImageAssetById(imageId);
  if (!image) {
    throw notFound("Image not found");
  }

  const variants = await findImageVariants(imageId);
  return variants.map(toVariantDto);
}

export async function deleteImage(imageId) {
  const image = await findImageAssetById(imageId);
  if (!image) {
    throw notFound("Image not found");
  }

  await deleteImageAssetWithVariants(imageId);
  return { message: "Image deleted" };
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/image/ImageService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.image

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode

data class ImageUploadInput(
    val uploadedByUserId: Int?,
    val originalFilename: String,
    val mimeType: String,
    val size: Int,
    val category: String,
)

class ImageService(
    private val repository: ImageRepository,
) {
    fun upload(input: ImageUploadInput): StubResponse {
        val saved = storeImage(input)
        return StubResponse(status = HttpStatusCode.Created, data = assetPayload(saved))
    }

    fun uploadLegacy(input: ImageUploadInput): StubResponse {
        val saved = storeImage(input)
        return StubResponse(status = HttpStatusCode.Created, data = mapOf("url" to saved.originalUrl))
    }

    fun variants(imageId: Int): StubResponse {
        requireImage(imageId)
        return StubResponse(data = repository.listVariants(imageId).map(::variantPayload))
    }

    fun delete(imageId: Int): StubResponse {
        if (!repository.deleteImage(imageId)) {
            throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "이미지를 찾을 수 없습니다.")
        }
        return StubResponse(data = mapOf("message" to "이미지가 삭제되었습니다."))
    }

    fun storeImage(input: ImageUploadInput): ImageAssetRecord {
        validateInput(input)
        return repository.uploadImage(
            ImageUploadCommand(
                uploadedByUserId = input.uploadedByUserId,
                originalFilename = sanitizeFilename(input.originalFilename),
                mimeType = input.mimeType.lowercase(),
                size = input.size,
                category = normalizeCategory(input.category),
            ),
        )
    }

    fun preferredUrl(asset: ImageAssetRecord): String =
        asset.variants.firstOrNull { it.type == "LARGE" }?.url
            ?: asset.variants.firstOrNull { it.type == "MEDIUM" }?.url
            ?: asset.variants.firstOrNull()?.url
            ?: asset.originalUrl

    fun preferredVariantId(asset: ImageAssetRecord): Int? =
        asset.variants.firstOrNull { it.type == "LARGE" }?.id
            ?: asset.variants.firstOrNull { it.type == "MEDIUM" }?.id
            ?: asset.variants.firstOrNull()?.id

    private fun requireImage(imageId: Int): ImageAssetRecord =
        repository.findImageById(imageId)
            ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "이미지를 찾을 수 없습니다.")

    private fun validateInput(input: ImageUploadInput) {
        if (input.originalFilename.isBlank()) {
            throw PbShopException(HttpStatusCode.BadRequest, "FILE_UPLOAD_FAILED", "업로드할 파일이 필요합니다.")
        }
        if (input.size <= 0) {
            throw PbShopException(HttpStatusCode.BadRequest, "FILE_UPLOAD_FAILED", "업로드할 파일이 필요합니다.")
        }
        if (input.size > MAX_FILE_SIZE) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "최대 10MB 파일만 업로드할 수 있습니다.")
        }
        if (!ALLOWED_MIME_TYPES.contains(input.mimeType.lowercase())) {
            throw PbShopException(HttpStatusCode.BadRequest, "FILE_TYPE_NOT_ALLOWED", "허용되지 않는 이미지 형식입니다.")
        }
        normalizeCategory(input.category)
    }

    private fun sanitizeFilename(filename: String): String =
        filename.trim().ifBlank { "image.bin" }.replace("\\s+".toRegex(), "-")

    private fun normalizeCategory(category: String): String {
        val normalized = category.trim().lowercase()
        if (!ALLOWED_CATEGORIES.contains(normalized)) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "지원하지 않는 이미지 카테고리입니다.")
        }
        return normalized
    }

    private fun assetPayload(asset: ImageAssetRecord): Map<String, Any?> =
        mapOf(
            "id" to asset.id,
            "originalUrl" to asset.originalUrl,
            "variants" to asset.variants.map(::variantPayload),
            "processingStatus" to asset.processingStatus,
        )

    private fun variantPayload(variant: ImageVariantRecord): Map<String, Any?> =
        mapOf(
            "id" to variant.id,
            "imageId" to variant.imageId,
            "type" to variant.type,
            "url" to variant.url,
            "width" to variant.width,
            "height" to variant.height,
            "format" to variant.format,
            "size" to variant.size,
            "createdAt" to variant.createdAt.toString(),
        )

    companion object {
        private const val MAX_FILE_SIZE = 10 * 1024 * 1024
        private val ALLOWED_MIME_TYPES = setOf("image/jpeg", "image/png", "image/webp", "image/gif")
        private val ALLOWED_CATEGORIES = setOf("product", "community", "support", "seller", "profile")
    }
}
